Docker local development set up

reyvababtista opened the pull request:
This pull request contains Docker configuration scripts (Dockerfile(s) and Docker Compose yml) for developing Beiwe locally without the need to deploy to Amazon Web Services.

There are 7 containers that will be built using the compose file `dev.docker-compose.yml`:
1. `storage` and `storage-client` for AWS S3 storage replacement using Minio.
2. `postgres` as the database server.
3. `web` is the Beiwe backend server.
4. `nginx` as the web server.
5. `rabbitmq` for message queue service.
6. `forest` as the data analysis service.

To run, execute these commands sequentially from the root directory. Make sure you fill out all the required environment variables inside `.envs/.env.dev` beforehand.

- Build the base image containing all of Beiwe's dependency requirements.
`docker build -f ./docker_config/backend/dev.base.Dockerfile -t beiwe-server-dev-base .`

- Build all the containers.
`docker compose -f docker_config/dev.docker-compose.yml --env-file ./.envs/.env.dev up --build`

- To kill the process, use `Cmd + C` or execute this command
`docker compose -f docker_config/dev.docker-compose.yml down -v`

TODO
- [x] Refine documentation.

biblicabeebli commented:
Questions:
- Is the `.envs` folder name required by docker, or is this configurable? `.envs` is maximally generic. (I'm refactoring 10 years of grunge and false expectations to clean up the base of the repo folder, it is probably time for the server backend code to go into a subfolder, or maybe even rethink the monorepo)
- It looks like there are some changes to not use sudo in various locations - can you confirm and slightly elaborate on this topic to confirm that I understand some Docker details:
  - While Docker itself is (should?) run under user privileges, inside the container by default is run as the root user.
  - From a security perspective this is acceptable because (in principle) any potential security issue within the container is limited to the container.
  - (Does the container include apt?)
- What is the development loop when updating site or backend code on this setup? what's the sync/update operation? It's been `python manage.py runserver` + postgres for ages. (There is code integrated into the codebase that fakes the existence of celery task queues when running locally and when running tests.)
- Where does the word use of the word "entrypoint" come from over in the RabbitMQ setup? What even is an entrypoint into a queue?  is it just... the name of the queue that you target?  (RabbitMQ is just part of our tech stack due to Celery. I have never understood why we need a non-database solution for hosting an atomic queue.)


Obscure Compatability Thing to be aware of:
https://github.com/onnela-lab/beiwe-backend/blob/e62a962137dd55899f4b98262feff91f65a9152d/constants/common_constants.py#L12-L15

We use the above variables in a couple places, most critically to bypass requiring Celery when running tests, but also it affects some error propagation to Sentry.

<hr>


Pushed files - I would like to deduplicate and instead parameterize the new files, because it is a maintenance burden otherwise.
- `bash_profile_docker.sh` needs some attention regardless because there are still uses of sudo here, as well as more hardcoded paths in some aliases, like `log`
  - [I don't know that ~sshing into this docker container is... relevant? I suspect you just copied and updated this one due to a file path error in the sneakily-keep-this-bash-profile-up-to-date code, which may itself be irrelevant.]
  - `$HOME` exists as a pre-filled variable to slot into file paths
  - to make logic for including sudo's into the aliases that need them: test against that $HOME variable, e.g. is it `/root` or `/home/ubuntu`, turn that into a $SUDO variable to inject into aliases that need it.  Careful on anything declared with single-quote strings.
- `cron_manager_docker.txt` - I just checked, the `/usr/bin/chronic` file path is also correct on ubuntu servers, so get rid of the docker-specific file and update both the worker and manager to target `/usr/bin/chronic`  (Oh, root doesn't import `/usr/bin` into its environment by default, that's what the `/usr/` bit implies.)
- `install_celery_worker.sh`
  - instead of removing those sudos, let's just change line 172 of the launch_script.py from `run(f'{REMOTE_INSTALL_CELERY_WORKER} >> {LOG_FILE}')` to `sudo(f'{REMOTE_INSTALL_CELERY_WORKER} >> {LOG_FILE}')` Make that change, I will test that soon anyway so just go ahead with that change.

<hr>

`config/django_settings.py`:
- please document the `S3_ENDPOINT` parameter as optional, when to include it, and it's purpose. (that file is the de-facto source of truth on all configuration details, it is actually linked directly from other documentation.)

<hr>

`docker_config/forest/dev.Dockerfile`
- This is misnamed. It should be called processing or data processing or maybe Celery, or even split into forest, push notifications, processing, and .... other tasks.
- Celery is the distributed task runner, we use it to send push notifications,, clean up and separating uploaded data into hourly files, run forest tasks, and run other system tasks that I need to put together, for instance periodically uploading access logs to s3 over on the ubuntu servers (a point of possible incompatibility here that I will have to think about, hm).
- For now just rename it to ... Celery. uhg. but that is the least-bad name and then the existence of RabbitMQ is explained just by looking at the container names.

<hr>

Whitespace and indentations:
- Your IDE looks configured to ~automatically trim trailing whitespace. While this has no functional effect whitespace characters are used extensively throughout the codebase to maintain legibility of complex sections. I know it is annoying, but I need you to turn on whitespace diffing in your git config to catch this and to revert the whitespace changes in `django_settings.py`.
  - (I have encountered annoyances where VSCode will format an entire file whenever format-on-paste is triggered.)
- The indentation fix in django settings to that TEMPLAITES declaration is correct though, woops.
- This project has a `.isort.cfg` file, your IDE applied an incorrect refactor to `libs/s3.py`

<hr>

Oddity in `docker_config/rabbitmq/entrypoint.sh`:
- `echo "*** Log in the WebUI at port 15672 (example: http://localhost:15672) ***") &` <~ run the print statement in the background?

<hr>

Let me know if you have any quibbles or questions.  Once this is mature I will make the effort to use it myself, or at least get the staging server to do so.

Also.

## It sure will be nice to have local S3 working for local development :D
So I'm excited.

reyvababtista commented:
> Are there any restrictions I need to know about names of files, placements of files, particularly critical exact orderings that can't be violated, or strings that must have exact values that are not otherwise documented in the content of this commit?
I think the restrictions would be the folder structure inside `docker_config`. If you are to change this structure, you will need to change every path declared in either `Dockerfile` or `*.docker-compose.yml`.

> First such question: does the docker folder itself need to be named as it is or could it possibly shift to `docker_management` to be in line with `cluster_management`
It can be changed. I will change it.

> I've made a branch `test-docker-onnela` that pulls in your work, its just for my own local diffing for example of your new cron file compared to the old manager file.
Got it.

> * Is the `.envs` folder name required by docker, or is this configurable? `.envs` is maximally generic. (I'm refactoring 10 years of grunge and false expectations to clean up the base of the repo folder, it is probably time for the server backend code to go into a subfolder, or maybe even rethink the monorepo)
This is configurable. The environment files were on the root folder originally, I moved it to `.envs` for easy configuration.

> * It looks like there are some changes to not use sudo in various locations - can you confirm and slightly elaborate on this topic to confirm that I understand some Docker details:
>   
>   * While Docker itself is (should?) run under user privileges, inside the container by default is run as the root user.
>   * From a security perspective this is acceptable because (in principle) any potential security issue within the container is limited to the container.


>   * (Does the container include apt?)
It depends on the Docker base image that you are using, to my knowledge any Debian based image will include `apt`.

> * What is the development loop when updating site or backend code on this setup? what's the sync/update operation? It's been `python manage.py runserver` + postgres for ages. (There is code integrated into the codebase that fakes the existence of celery task queues when running locally and when running tests.)
Pull in the changes from the root directory of this repository, then do `docker compose run ...`

> * Where does the word use of the word "entrypoint" come from over in the RabbitMQ setup? What even is an entrypoint into a queue?  is it just... the name of the queue that you target?  (RabbitMQ is just part of our tech stack due to Celery. I have never understood why we need a non-database solution for hosting an atomic queue.)
See this line: https://github.com/onnela-lab/beiwe-backend/pull/396/files#diff-f5ef0d60a3febdc2ef385db908bbf3fdd47bbd453f72f2c908841375ed59d710R15
If you are referring to `ENTRYPOINT` instruction, it is to run executables each time the container starts. If you are referring to `entrypoint.sh`, it is to create RabbitMQ username (see https://github.com/onnela-lab/beiwe-backend/blob/e62a962137dd55899f4b98262feff91f65a9152d/cluster_management/launch_script.py#L501), which we can only do so after RabbitMQ container is initialized successfully, hence the `ENTRYPOINT`.

> Obscure Compatability Thing to be aware of:
> 
> https://github.com/onnela-lab/beiwe-backend/blob/e62a962137dd55899f4b98262feff91f65a9152d/constants/common_constants.py#L12-L15
> 
> We use the above variables in a couple places, most critically to bypass requiring Celery when running tests, but also it affects some error propagation to Sentry.
I see, how should I implement this?

> Pushed files - I would like to deduplicate and instead parameterize the new files, because it is a maintenance burden otherwise.
> 
> * `bash_profile_docker.sh` needs some attention regardless because there are still uses of sudo here, as well as more hardcoded paths in some aliases, like `log`
>   
>   * [I don't know that ~sshing into this docker container is... relevant? I suspect you just copied and updated this one due to a file path error in the sneakily-keep-this-bash-profile-up-to-date code, which may itself be irrelevant.]
Yep, this was exactly what I did. Lol.

>   * `$HOME` exists as a pre-filled variable to slot into file paths
>   * to make logic for including sudo's into the aliases that need them: test against that $HOME variable, e.g. is it `/root` or `/home/ubuntu`, turn that into a $SUDO variable to inject into aliases that need it.  Careful on anything declared with single-quote strings.
> * `cron_manager_docker.txt` - I just checked, the `/usr/bin/chronic` file path is also correct on ubuntu servers, so get rid of the docker-specific file and update both the worker and manager to target `/usr/bin/chronic`  (Oh, root doesn't import `/usr/bin` into its environment by default, that's what the `/usr/` bit implies.)
Got it.

> * `install_celery_worker.sh`
>   
>   * instead of removing those sudos, let's just change line 172 of the launch_script.py from `run(f'{REMOTE_INSTALL_CELERY_WORKER} >> {LOG_FILE}')` to `sudo(f'{REMOTE_INSTALL_CELERY_WORKER} >> {LOG_FILE}')` Make that change, I will test that soon anyway so just go ahead with that change.
Great input. Will do.

> `config/django_settings.py`:
> 
> * please document the `S3_ENDPOINT` parameter as optional, when to include it, and it's purpose. (that file is the de-facto source of truth on all configuration details, it is actually linked directly from other documentation.)
Will do.

> `docker_config/forest/dev.Dockerfile`
> 
> * This is misnamed. It should be called processing or data processing or maybe Celery, or even split into forest, push notifications, processing, and .... other tasks.
> * Celery is the distributed task runner, we use it to send push notifications,, clean up and separating uploaded data into hourly files, run forest tasks, and run other system tasks that I need to put together, for instance periodically uploading access logs to s3 over on the ubuntu servers (a point of possible incompatibility here that I will have to think about, hm).
> * For now just rename it to ... Celery. uhg. but that is the least-bad name and then the existence of RabbitMQ is explained just by looking at the container names.
Got it.

> Whitespace and indentations:
> 
> * Your IDE looks configured to ~automatically trim trailing whitespace. While this has no functional effect whitespace characters are used extensively throughout the codebase to maintain legibility of complex sections. I know it is annoying, but I need you to turn on whitespace diffing in your git config to catch this and to revert the whitespace changes in `django_settings.py`.
>   
>   * (I have encountered annoyances where VSCode will format an entire file whenever format-on-paste is triggered.)
> * The indentation fix in django settings to that TEMPLAITES declaration is correct though, woops.
> * This project has a `.isort.cfg` file, your IDE applied an incorrect refactor to `libs/s3.py`
I will need to look into this first, I have never paid close attention to white space clean ups by IDEs.

> Oddity in `docker_config/rabbitmq/entrypoint.sh`:
> 
> * `echo "*** Log in the WebUI at port 15672 (example: http://localhost:15672) ***") &` <~ run the print statement in the background?
I think I copy-pasted this line from somewhere, I can't remember. I will remove this.

> Let me know if you have any quibbles or questions. Once this is mature I will make the effort to use it myself, or at least get the staging server to do so.
> 
> Also.
> 
> ## It sure will be nice to have local S3 working for local development :D
> So I'm excited.
Thanks! I am currently working on the production code for this. So this will be my first PR from two PRs in queue. The second one will remove Minio to use S3 and will include HTTPS certification using Certbot.
